File: Cpp/Codes/Practice/LeetCode/649 Dota2 Senate.cpp
#include <gtest\gtest.h>
using namespace std;

static string predictPartyVictory(string senate)
{	
	int banDelta = 0; // > 0 表示需要禁言D的个数，< 0 表示需要禁言R的个数
	int countOfR;
	int countOfD;

	do
	{	
		countOfR = 0;
		countOfD = 0;
		string current;
		for (size_t i = 0; i < senate.length(); ++i)
		{
			if (senate[i] == 'R')
			{
				if (banDelta >= 0)
				{
					current += "R";
					countOfR++;
				}

				banDelta++;
			}
			else
			{
				if (banDelta <= 0)
				{
					current += "D";
					countOfD++;
				}

				banDelta--;
			}
		}

		senate = current;

	} while (countOfR && countOfD);
	
	return (senate[0] == 'R') ? "Radiant" : "Dire";
}

TEST(LeetCode, tPredictPartyVictory)
{
	ASSERT_EQ(predictPartyVictory("RD"),"Radiant");
	ASSERT_EQ(predictPartyVictory("RDD"), "Dire");
	ASSERT_EQ(predictPartyVictory("RDDRDRDDDRDDRRRDR"), "Dire");
}

File: Python/Codes/Practice/LeetCode/649 Dota2 Senate.py

class Solution:
    def predictPartyVictory(self, senate):
        """
        :type senate: str
        :rtype: str
        """
        
        banSenate = ''        
        while len(set(senate)) > 1:
            
            remainSenate = ''
            for val in senate:
                if banSenate and banSenate[0] == val:             
                    banSenate = banSenate[1:]
                else:
                    if val == 'R':
                        banSenate += 'D'
                    else:
                        banSenate += 'R'

                    remainSenate += val
            
            senate = remainSenate
           
        return {'D':"Dire",'R':"Radiant"}[senate[0]]
            

if __name__ == '__main__':

    solution = Solution()
    print(solution.predictPartyVictory('RD'))
    print(solution.predictPartyVictory('RDD'))
    print(solution.predictPartyVictory('RDDRDRDDDRDDRRRDR'))